File: ksheermitra/lib/models/product.dart
class Product {
  final String id;
  final String name;
  final String? description;
  final String? imageUrl;
  final String unit;
  final double pricePerUnit;
  final bool isActive;
  final int stock;

  Product({
    required this.id,
    required this.name,
    this.description,
    this.imageUrl,
    required this.unit,
    required this.pricePerUnit,
    required this.isActive,
    required this.stock,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id']?.toString() ?? '',
      name: json['name']?.toString() ?? 'Unknown Product',
      description: json['description']?.toString(),
      imageUrl: json['imageUrl']?.toString(),
      unit: json['unit']?.toString() ?? 'unit',
      pricePerUnit: json['pricePerUnit'] != null
          ? double.tryParse(json['pricePerUnit'].toString()) ?? 0.0
          : 0.0,
      isActive: json['isActive'] ?? true,
      stock: json['stock'] != null
          ? int.tryParse(json['stock'].toString()) ?? 0
          : 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'imageUrl': imageUrl,
      'unit': unit,
      'pricePerUnit': pricePerUnit,
      'isActive': isActive,
      'stock': stock,
    };
  }
}

File: backend/src/models/Product.js
module.exports = (sequelize, DataTypes) => {
  const Product = sequelize.define('Product', {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true
    },
    name: {
      type: DataTypes.STRING,
      allowNull: false,
      unique: true,
      validate: {
        notEmpty: true,
        len: [2, 100]
      }
    },
    description: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    imageUrl: {
      type: DataTypes.STRING,
      allowNull: true
    },
    unit: {
      type: DataTypes.STRING(20),
      allowNull: false,
      defaultValue: 'liter',
      validate: {
        isIn: [['liter', 'ml', 'kg', 'gm', 'piece']]
      }
    },
    pricePerUnit: {
      type: DataTypes.DECIMAL(10, 2),
      allowNull: false,
      validate: {
        min: 0,
        isDecimal: true
      }
    },
    isActive: {
      type: DataTypes.BOOLEAN,
      defaultValue: true
    },
    stock: {
      type: DataTypes.INTEGER,
      defaultValue: 0,
      validate: {
        min: 0
      }
    }
  }, {
    tableName: 'Products',
    timestamps: true,
    paranoid: true,
    indexes: [
      {
        unique: true,
        fields: ['name']
      },
      {
        fields: ['isActive']
      }
    ]
  });

  Product.associate = (models) => {
    Product.hasMany(models.Subscription, {
      foreignKey: 'productId',
      as: 'subscriptions'
    });

    Product.hasMany(models.Delivery, {
      foreignKey: 'productId',
      as: 'deliveries'
    });
  };

  return Product;
};
